File: Data Structures & Algorithms/binary-search/submission-0.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        return helper(nums, target, 0, nums.size() - 1);
    }

    int helper(vector<int>& nums, int target, int l, int r) {
        if (r < l) return -1;

        int n = r - l;
        int mid = l + (r - l) / 2;
        if (nums[mid] == target) {
            return mid;
        }

        if (target > nums[mid]) {
            l = mid + 1;
        } else {
            r = mid - 1;
        }

        return helper(nums, target, l, r);
    }
};

File: Data Structures & Algorithms/count-paths/submission-0.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = {}
        
        def helper(x, y, m, n):
            if x >= m or y >= n:
                return 0
            if x == m - 1 and y == n - 1:
                return 1

            if (x, y) in dp:
                return dp[(x, y)]
            
            paths = helper(x+1, y, m, n) + helper(x, y+1, m, n)
            dp[(x, y)] = paths
            return paths
        
        return helper(0, 0, m, n)